### PR TITLE
xds: implement new CDS LB policy for supporting aggregate clusters

### PR DESCRIPTION
**Pull request by voidzcy**
Implementation for (new) CDS LB policy that supports cluster discovery for aggregate clusters.

The new implementation takes takes the same LB config as before, which contains a single cluster name. But this cluster can be representing an aggregate cluster, which is a composition for a tree-structured cluster hierarchy (precisely speaking, it can be a directed graph with no cycles. That is, the hierarchy can contain diamond, where a cluster belongs to more than one aggregate clusters). Therefore, the output (aka, LB config for its child LB policy, which is `ClustserResolverLoadBalancer`) contains a list of cluster configurations (aka, `DiscoveryMechanism`).

When receiving its LB config, the CDS LB policy starts watching on the CDS resource (the root) using the cluster name from the LB config. Upon receiving a resource update, which is the configuration for the cluster, it can have the following two cases:

- The cluster is an EDS/Logical DNS cluster: we've reached the end of cluster discovery for this cluster, output it as an DiscoveryMechanism entry.
- The cluster is an aggregate cluster: the update contains a list of cluster names (in order) that composites this cluster. Then we start one CDS watcher for each of the contained clusters.

The discovery process happens recursively for a tree-hierarchy aggregate cluster until all (leaf) clusters have been discovered. Example and details can be seen in go/grpc-xds-directpath-c2p-fallback.

A few implementation assumptions for cluster resources:
- Clusters are uniquely identified by its name and a cluster can never change its _intrinsic property_ (e.g., an EDS cluster will never become a Logical DNS or aggregate cluster).
- The composition of an aggregate cluster is allowed to be changed (e.g., aggregate cluster containing [A, B, C] can change to contain [A, B] or [D]).

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+                  watcher.shutdown();
+                }
+              }
+              childClusterStates = newChildStates;
+            } else if (update.clusterType == ClusterType.EDS) {
+              isLeaf = true;
+              EdsClusterConfig clusterConfig = (EdsClusterConfig) update.clusterConfig;
+              logger.log(XdsLogLevel.INFO, "EDS cluster {0}, edsServiceName: {1}",
+                  update.clusterName, clusterConfig.edsServiceName);
+              logger.log(XdsLogLevel.DEBUG, "Cluster config: {0}", clusterConfig);
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java:312]: also revoke child clusters if `if (childClusterStates != null)`, and then set `childClusterStates = null`?

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java:312]: I don't think we need to consider such a weird use case: a EDS/Logical DNS cluster becomes an aggregate cluster (or vice versa).

**Review by dapengzhang0 — APPROVED** [review]
LGTM. When is it ready to migrate `CdsLoadBalancerProvider`?

**Comment by voidzcy** [comment]
> LGTM. When is it ready to migrate `CdsLoadBalancerProvider`?

It would be right for the next PR, and I will at least try it out to ensure interop tests will not be broken.

File: xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import javax.annotation.Nullable;
+
+/**
+ * Load balancer for cds_experimental LB policy. One instance per top-level cluster.
+ * The top-level cluster may be a plain EDS/logical-DNS cluster or an aggregate cluster that
+ * formed by a group of sub-clusters in a tree hierarchy.
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java]: Fixed.

File: xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+      xdsClientPool.returnObject(xdsClient);
+    }
+  }
+
+  /**
+   * The state of a CDS working session of {@link CdsLoadBalancer2}. Created and started when
+   * receiving the CDS LB policy config with the top-level cluster name.
+   */
+  private final class CdsLbState {
+    private ClusterState root;
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java]: Fixed.

**Pull request merged**